### PR TITLE
virt-operator: replace hardcoded patch operations 

### PR DESCRIPTION
**Pull request by alicefr**
Use the patch package to replace the hardcoded patch operations. This improves the readability and make the code more cohesive.

### What this PR does
Before this PR:
Hardcoded operation in virt-operator/apply

After this PR:
Replaced all the patch operation with the patch packge

Fixes #
Partially fix: https://github.com/kubevirt/kubevirt/issues/11654

### Release note

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-operator/resource/apply/reconcile.go [hunk 1]
```diff
@@ -291,40 +284,17 @@ func generatePatchBytes(ops []string) []byte {
 	return controller.GeneratePatchBytes(ops)
```
Comment (fossedihelm) [inline pkg/virt-operator/resource/apply/reconcile.go:284]: we can remove the `generatePatchBytes` function now :)

File: pkg/virt-operator/resource/apply/admissionregistration.go [hunk 2]
```diff
@@ -219,27 +220,15 @@ func (r *Reconciler) createOrUpdateValidatingWebhookConfiguration(webhook *admis
...
-	ops := []string{
-		fmt.Sprintf(testGenerationJSONPatchTemplate, cachedWebhook.ObjectMeta.Generation),
-	}
-
+	ops := []patch.PatchOption{patch.WithTest("/metadata/generation", cachedWebhook.ObjectMeta.Generation)}
 	// Add Labels and Annotations Patches
-	labelAnnotationPatch, err := createLabelsAndAnnotationsPatch(&webhook.ObjectMeta)
+	ops = append(ops, createLabelsAndAnnotationsPatch(&webhook.ObjectMeta)...)
+	ops = append(ops, patch.WithReplace("/webhooks", webhook.Webhooks))
+	patch, err := patch.New(ops...).GeneratePayload()
```
Comment (fossedihelm) [inline pkg/virt-operator/resource/apply/admissionregistration.go:227]: How about:
```
	// Patch if old version
	patchSet := patch.New(
		patch.WithTest("/metadata/generation", cachedWebhook.ObjectMeta.Generation),
	)
	// Add Labels and Annotations Patches
	patchSet.AddOption(createLabelsAndAnnotationsPatch(&webhook.ObjectMeta)...)
	patchSet.AddOption(patch.WithReplace("/webhooks", webhook.Webhooks))
	patchBytes, err := patchSet.GeneratePayload()
	if err != nil {
		return err
	}
	webhook, err = r.patchValidatingWebhookConfiguration(webhook, patchBytes)
```
?
Same for other occurrences below

File: pkg/virt-operator/resource/apply/apiservices.go [hunk 3]
```diff
@@ -77,12 +79,13 @@ func (r *Reconciler) createOrUpdateAPIService(apiService *apiregv1.APIService, c
 		return err
 	}
 
-	ops, err := getPatchWithObjectMetaAndSpec([]string{}, &apiService.ObjectMeta, spec)
+	ops := getPatchWithObjectMetaAndSpec([]patch.PatchOption{}, &apiService.ObjectMeta, spec)
+	patch, err := patch.New(ops...).GeneratePayload()
```
Comment (fossedihelm) [inline pkg/virt-operator/resource/apply/apiservices.go:83]: we can inline this

File: pkg/virt-operator/resource/apply/crds.go [hunk 4]
```diff
@@ -113,15 +112,11 @@ func (r *Reconciler) createOrUpdateCrd(crd *extv1.CustomResourceDefinition) erro
 		}
 	}
 	// Add Labels and Annotations Patches
-	var ops []string
-	labelAnnotationPatch, err := createLabelsAndAnnotationsPatch(&crd.ObjectMeta)
+	ops := createLabelsAndAnnotationsPatch(&crd.ObjectMeta)
+	crd, err := patchCRD(client, crd, ops)
```
Comment (fossedihelm) [inline pkg/virt-operator/resource/apply/crds.go:116]: inline?

File: pkg/virt-operator/resource/apply/core_test.go [hunk 5]
```diff
@@ -395,24 +396,19 @@ var _ = Describe("Apply", func() {
...
-				hasSubstring := func(ops []string, substring string) bool {
-					for _, op := range ops {
-						if strings.Contains(op, substring) {
-							return true
-						}
-					}
-					return false
+				if !expectLabelsAnnotationsPatch && !expectSpecPatch {
+					Expect(ops).To(BeEmpty())
+				} else {
```
Comment (fossedihelm) [inline pkg/virt-operator/resource/apply/core_test.go:401]: empty `else`?

File: pkg/virt-operator/resource/apply/core_test.go [hunk 6]
```diff
@@ -395,24 +396,19 @@ var _ = Describe("Apply", func() {
...
 				}
+				Expect(err).ToNot(HaveOccurred())
```
Comment (fossedihelm) [inline pkg/virt-operator/resource/apply/core_test.go:403]: I'd expect this to be the first instruction after `ops, err := generateServicePatch(cachedService, targetService)`

File: pkg/virt-operator/resource/apply/core_test.go [hunk 7]
```diff
@@ -395,24 +396,19 @@ var _ = Describe("Apply", func() {
...
 
 				if expectLabelsAnnotationsPatch {
-					Expect(hasSubstring(ops, "/metadata/labels")).To(BeTrue())
-					Expect(hasSubstring(ops, "/metadata/annotations")).To(BeTrue())
+					Expect(strings.Contains(string(ops), "/metadata/labels")).To(BeTrue())
+					Expect(strings.Contains(string(ops), "/metadata/annotations")).To(BeTrue())
```
Comment (fossedihelm) [inline pkg/virt-operator/resource/apply/core_test.go:407]:
```suggestion
					Expect(string(ops)).To(ContainSubstring("/metadata/labels"))
					Expect(string(ops)).To(ContainSubstring("/metadata/annotations"))
```
same for the `expectSpecPatch` below

**Review by fossedihelm — COMMENTED** [review]
Thank you @alicefr!
Some comments below, maybe there are some leftovers :)

File: pkg/virt-operator/resource/apply/apps.go [hunk 8]
```diff
@@ -386,18 +373,12 @@ func (r *Reconciler) syncPodDisruptionBudgetForDeployment(deployment *appsv1.Dep
...
 	}
 
-	// Add Spec Patch
-	newSpec, err := json.Marshal(podDisruptionBudget.Spec)
-	if err != nil {
-		return err
-	}
-
-	ops, err := getPatchWithObjectMetaAndSpec([]string{}, &podDisruptionBudget.ObjectMeta, newSpec)
+	patch, err := patch.New(getPatchWithObjectMetaAndSpec([]patch.PatchOption{}, &podDisruptionBudget.ObjectMeta, podDisruptionBudget.Spec)...).GeneratePayload()
```
Comment (fossedihelm) [inline pkg/virt-operator/resource/apply/apps.go:376]: nit: ```patchBytes, err := ...```
warning: `Variable 'patch' collides with imported package name `

File: pkg/virt-operator/resource/apply/core.go [hunk 9]
```diff
@@ -134,7 +135,7 @@ func (r *Reconciler) createOrUpdateService(service *corev1.Service) (bool, error
 		return false, nil
 	}
 
-	_, err = core.Services(service.Namespace).Patch(context.Background(), service.Name, types.JSONPatchType, generatePatchBytes(patchOps), metav1.PatchOptions{})
+	_, err = core.Services(service.Namespace).Patch(context.Background(), service.Name, types.JSONPatchType, patchOps, metav1.PatchOptions{})
```
Comment (fossedihelm) [inline pkg/virt-operator/resource/apply/core.go:138]: nit: `patchOps` -> `patchBytes`

File: pkg/virt-operator/resource/apply/core.go [hunk 10]
```diff
@@ -270,7 +271,7 @@ func (r *Reconciler) createOrUpdateCertificateSecret(queue workqueue.RateLimitin
 		return nil, err
 	}
 
-	_, err = r.clientset.CoreV1().Secrets(secret.Namespace).Patch(context.Background(), secret.Name, types.JSONPatchType, generatePatchBytes(ops), metav1.PatchOptions{})
+	_, err = r.clientset.CoreV1().Secrets(secret.Namespace).Patch(context.Background(), secret.Name, types.JSONPatchType, ops, metav1.PatchOptions{})
```
Comment (fossedihelm) [inline pkg/virt-operator/resource/apply/core.go:274]: same as above

File: pkg/virt-operator/resource/apply/apps.go (hunk 8 above)
Comment (fossedihelm) [reply pkg/virt-operator/resource/apply/apps.go:376]: There are other occurrences :)

File: pkg/virt-operator/resource/apply/core_test.go [hunk 11]
```diff
@@ -396,23 +396,17 @@ var _ = Describe("Apply", func() {
...
-					return false
+				if !expectLabelsAnnotationsPatch && !expectSpecPatch {
+					Expect(ops).To(BeEmpty())
 				}
 
 				if expectLabelsAnnotationsPatch {
-					Expect(hasSubstring(ops, "/metadata/labels")).To(BeTrue())
-					Expect(hasSubstring(ops, "/metadata/annotations")).To(BeTrue())
+					Expect(string(ops)).To(ContainSubstring("/metadata/labels"))
+					Expect(string(ops)).To(ContainSubstring("/metadata/annotations"))
```
Comment (fossedihelm) [inline pkg/virt-operator/resource/apply/core_test.go:405]: :+1:  :partying_face:

**Review by fossedihelm — COMMENTED** [review]
Thank you @alicefr!
Overall looks good! There is a change in the behavior regarding the metadata patch or some resources. Was it intentional?
Thanks

File: pkg/virt-operator/resource/apply/admissionregistration.go [hunk 12]
```diff
@@ -412,20 +369,25 @@ func (r *Reconciler) createOrUpdateValidatingAdmissionPolicyBinding(validatingAd
...
 	cachedValidatingAdmissionPolicyBinding := obj.(*admissionregistrationv1.ValidatingAdmissionPolicyBinding)
 
-	patchOps, err := generateValidatingAdmissionPolicyBindingPatch(cachedValidatingAdmissionPolicyBinding, validatingAdmissionPolicyBinding)
-	if err != nil {
-		return fmt.Errorf("unable to generate validatingAdmissionPolicyBinding patch operations for %+v: %v", validatingAdmissionPolicyBinding, err)
+	patchSet := patch.New()
+	if !equality.Semantic.DeepEqual(cachedValidatingAdmissionPolicyBinding.Spec, validatingAdmissionPolicyBinding.Spec) {
+		patchSet.AddOption(getObjectMetaPatch(validatingAdmissionPolicyBinding.ObjectMeta,
+			cachedValidatingAdmissionPolicyBinding.ObjectMeta)...)
+		patchSet.AddOption(patch.WithReplace("/spec", validatingAdmissionPolicyBinding.Spec))
```
[the remark this answers is not in the document]
Comment (alicefr) [reply pkg/virt-operator/resource/apply/admissionregistration.go:376]: You are, of course, right! This is probably also the reason  why the CI is failing

**Review by fossedihelm — COMMENTED** [review]
Some last cosmetic stuff. Thank you for your patience and work!

File: pkg/virt-operator/resource/apply/admissionregistration.go [hunk 13]
```diff
@@ -81,13 +82,13 @@ func convertV1beta1MutatingWebhookToV1(from *admissionregistrationv1beta1.Mutati
 	return webhookv1, nil
 }
 
-func (r *Reconciler) patchValidatingWebhookConfiguration(webhook *admissionregistrationv1.ValidatingWebhookConfiguration, ops []string) (patchedWebhook *admissionregistrationv1.ValidatingWebhookConfiguration, err error) {
+func (r *Reconciler) patchValidatingWebhookConfiguration(webhook *admissionregistrationv1.ValidatingWebhookConfiguration, ops []byte) (patchedWebhook *admissionregistrationv1.ValidatingWebhookConfiguration, err error) {
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-operator/resource/apply/admissionregistration.go:85]: This is the only one missed :)

**Review by fossedihelm — COMMENTED** [review]
Just one miss

**Comment by alicefr** [comment]
@fossedihelm yes, sorry. Now, it should be fine.

**Comment by fossedihelm** [comment]
Thank you @alicefr
/lgtm

**Comment by sarahbx** [comment]
Please reduce the code duplication to pass the SonarCloud gate.
https://sonarcloud.io/component_measures?metric=new_duplicated_lines_density&selected=kubevirt_kubevirt%3Apkg%2Fvirt-operator%2Fresource%2Fapply%2Fadmissionregistration.go&pullRequest=11976&id=kubevirt_kubevirt

**Comment by alicefr** [comment]
> Please reduce the code duplication to pass the SonarCloud gate. https://sonarcloud.io/component_measures?metric=new_duplicated_lines_density&selected=kubevirt_kubevirt%3Apkg%2Fvirt-operator%2Fresource%2Fapply%2Fadmissionregistration.go&pullRequest=11976&id=kubevirt_kubevirt

This PR replaces hardcoded operations with a package, I would avoid to make it more complicated by refactoring the code. This could be done in a follow up PR. @fossedihelm wdyt?

**Comment by fossedihelm** [comment]
To me, it's ok to address it in a follow-up PR but I don't know how can we skip the sonar cloud. tide will block this PR to gets merged.

**Comment by alicefr** [comment]
> To me, it's ok to address it in a follow-up PR but I don't know how can we skip the sonar cloud. tide will block this PR to gets merged.

Is it mandatory?

**Comment by fossedihelm** [comment]
As far as I know yes

**Comment by xpivarc** [comment]
> > To me, it's ok to address it in a follow-up PR but I don't know how can we skip the sonar cloud. tide will block this PR to gets merged.
> 
> Is it mandatory?

@alicefr 
Unfortunately there is no way to override.

**Comment by alicefr** [comment]
@xpivarc @fossedihelm it seems now happy with the current changes

File: pkg/virt-operator/resource/apply/core_test.go [hunk 14]
```diff
@@ -62,8 +61,9 @@ var _ = Describe("Apply", func() {
 			service := &corev1.Service{}
 			service.Spec.Type = corev1.ServiceTypeClusterIP
 			service.Spec.ClusterIP = ""
-
-			Expect(generateServicePatch(cachedService, service)).To(BeEmpty())
+			patch, err := generateServicePatch(cachedService, service)
```
Comment (xpivarc) [inline pkg/virt-operator/resource/apply/core_test.go:64]: This should still work.

**Review by xpivarc — COMMENTED** [review]
/approve

File: pkg/virt-operator/resource/apply/core_test.go (hunk 14 above)
Comment (alicefr) [reply pkg/virt-operator/resource/apply/core_test.go]: Right, thanks!

**Review by fossedihelm — APPROVED** [review]
Thank you!

**Pull request merged**